runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/util/StringValuesMap.kt

```diff
@@ -64,8 +64,10 @@ internal open class StringValuesMapImpl(
     override val caseInsensitiveName: Boolean = false,
     initialValues: Map<String, List<String>> = emptyMap()
 ) : StringValuesMap {
-    protected val values: Map<String, List<String>> by lazy {
-        if (caseInsensitiveName) CaseInsensitiveMap<List<String>>().apply { putAll(initialValues) } else values.toMap()
+    protected val values: Map<String, List<String>> = run {
+        // Make a defensive copy so modifications to the initialValues don't mutate our internal copy
+        val copiedValues = initialValues.deepCopy()
```

Comment:
I'm afraid it is necessary. The type of the _parameter_ may be the immutable interface type `Map` but the actual _argument_ supplied is a mutable implementation. Modifications to that map in the builder will cause already-built instances to mutate unless copied.

Moreover, the map values passed in are mutable lists. They too need to be defensively copied to prevent post hoc mutation.

Reply:
Which argument where?